src/Components/Components/src/Routing/RouteEntry.cs

```diff
@@ -62,8 +74,22 @@ internal void Match(RouteContext context)
                 }
             }
 
-            context.Parameters = parameters;
-            context.Handler = Handler;
+            // We track the number of segments in the template that matched
+            // against this particular route then only select the route that
+            // matches the most number of segments on the route that was passed.
+            // This check is an exactness check that favors the more precise of
+            // two templates in the event that the following route table exists.
+            //  Route 1: /{anythingGoes}
+            //  Route 2: /users/{id:int}
+            // And the provided route is `/users/1`. We want to choose Route 2
+            // over Route 1.
```

Comment:
The precedence rules you're describing sound right to me. @pranavkm Do you know how closely this matches the server-side precedence rules?

Also since precedence is both subtle and important, would it be possible to add some further test cases showing how the matching behaves when there is more than one possible match?

Reply:
Good call. I'll add more tests here.

Reply:
The precedence rule described in the comment sounds right. Routing also prefers literal segments to constraints. If If `/foo/bar` and `/foo/{x?}` match, the literal segment is preferred. This also applies to non-optional constraints, for instance the literal `/foo/bar` would be preferred over`/foo/{x}`.

Reply:
Makes sense. There were some existing test cases in the route tests that were validating this that still pass. I also went ahead and added a few more tests for this precedence logic.